Connector for IBM DB2

n-riesco opened the pull request:
This PR implements a connector for IBM DB2 using the npm native module `ibm_db`.

I've added the scripts `npm run fix:ibm_db` and `npm run postinstall` to ease build the electron app.

Here are the instructions for Linux and Mac to build the electron app with this connector:
```
git clone https://github.com/n-riesco/plotly-database-connector.git
cd plotly-database-connector
npm install
npm run fix:ibm_db
npm run postinstall
rm -f dist/*
npm run build
npm run start
```

n-riesco committed:
app: Fix logo overflow

* Updated `DialectSelector.css` so that logos don't overflow buttons.

n-riesco committed:
Add native module ibm_db

* Added `ibm_db` to the list of dependencies.

* Added `ibm_db` to the list of externals in `webpack.config.base.js`.

* Added the script `npm run postinstall` that needs to be run after
  `npm install` to ensure all native modules are linked against
  electron.

* Added the script `npm run fix:ibm_db` that needs to be run in Linux
  and MacOS before `npm run postinstall` to fix the file permissions of
  the `ibm_db` module.

n-riesco committed:
ibmdb: Initial commit of a connector for IBM DB2

chriddyp committed:
Security update node.js to LTS

n-riesco committed:
ibmdb: Update logo

* Use green background.

* Remove IBM to be able to use a larger font.

n-riesco committed:
ibmdb: Name tables using the schema name

n-riesco committed:
ibmdb: Add Dockerfile

* Added Dockerfile to set up an IBM DB2 Express-C instance with an admin
  user (db2inst1), a SELECT-only user (db2user1) and a database (sample)
  listening on port 50000.

* To build the docker image run:

  DB2INST1_PASSWORD=pass DB2USER1_PASSWORD=pass npm run docker:db2:build

* To start the instance, run:

  npm run docker:db2:start

* See license and further details in `test/docker/ibmdb2`.

chriddyp commented:
💃 - looks good to me! once we get a sample DB running we can add some acceptance tests as well

n-riesco committed:
ibmdb: Remove docker container when it exits

n-riesco committed:
Merge branch 'master' into db2

n-riesco committed:
ibmdb: Lock at version ibm_db@2.0

* ibm_db@2.1.0 fixes the clidriver permission issue. See:
  https://github.com/ibmdb/node-ibm_db/pull/284

* ibm_db@2.1.0 throws the following error on `npm run start`:

user@host:~/github/plotly-database-connector$ npm run start

> plotly-database-connector@2.0.2 start /home/user/github/plotly-database-connector
> cross-env NODE_ENV=production electron ./

App threw an error during load
Error: libdb2.so.1: cannot open shared object file: No such file or directory
    at process.module.(anonymous function) [as dlopen] (ELECTRON_ASAR.js:173:20)
    at Object.Module._extensions..node (module.js:598:18)
    at Object.module.(anonymous function) [as .node] (ELECTRON_ASAR.js:173:20)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at bindings (/home/user/github/plotly-database-connector/node_modules/bindings/bindings.js:76:44)
    at Object.<anonymous> (/home/user/github/plotly-database-connector/node_modules/ibm_db/lib/odbc.js:27:31)
    at Object.<anonymous> (/home/user/github/plotly-database-connector/node_modules/ibm_db/lib/odbc.js:1761:3)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.e.__esModule.default (/home/user/github/plotly-database-connector/backend/webpack:/external "ibm_db":1:1)
    at t (/home/user/github/plotly-database-connector/backend/webpack:/webpack/bootstrap 1020c4d3a5e21580f839:19:1)
    at Object.<anonymous> (/home/user/github/plotly-database-connector/backend/webpack:/backend/persistent/datastores/ibmdb2.js:1:15)
    at t (/home/user/github/plotly-database-connector/backend/webpack:/webpack/bootstrap 1020c4d3a5e21580f839:19:1)

n-riesco committed:
ibmdb: Add unit tests

n-riesco committed:
ibmdb: Update scripts to rebuild native modules

* Removed npm postinstall, because unit tests are run in node.

* Added `npm run install:modules:node` to rebuild the native modules
  before running the unit tests.

* Added `npm run install:modules:electron` to rebuild the native modules
  before building the app and running `npm start`.

* Note that `script/install-modules.js` invokes the global `npm`. The
  local `npm` fails to build the native module `ibm_db`.

n-riesco committed:
ibmdb: Connect using PROTOCOL=TCPIP

n-riesco committed:
ibmdb: Update address of test server

n-riesco committed:
ibmdb: Add credentials for DB2 to heroku app

n-riesco committed:
ibmdb: Workaround fsevents install with electron

* See https://github.com/strongloop/fsevents/issues/131

n-riesco committed:
ibmdb: Rebuild native modules before test-e2e

bpostlethwaite committed:
inject plotly dataset data into db named 'plotly'

n-riesco committed:
Merge pull request #3 from plotly/usePlotlyDataSetsInDB2

ibmdb: Update docker image

* Injected plotly dataset into db named 'plotly'.

n-riesco committed:
Merge branch 'db2' of https://github.com/n-riesco/plotly-database-connector into db2

n-riesco committed:
ibmdb: Update docker image

* Skip header when importing the CSV file.

n-riesco committed:
ibmdb: Update database name

n-riesco committed:
ibmdb: Update tests

* Use table ALCOHOL_CONSUMPTION_BY_COUNTRY_2010.

n-riesco committed:
ibmdb: Fix native modules before running test-e2e

* Do not use local `npm` to rebuild native modules.

* Increased default timeout to 10s, so that tests are less likely to
  fail in slow machines and connections.

n-riesco committed:
circle: Do not cache node_modules

* Do not cache `node_modules`, otherwise it is not possible to rebuild
  the native module `ibm_db` before running `npm run test-e2e`. See
  `npm run test` for further details.

n-riesco committed:
circle: Do not cache node_modules

* Add flag to force removal.

n-riesco committed:
circle: Do not cache node_modules

* Disable cache of `node_modules` by moving `npm install` from
  `dependencies` to `test`.

n-riesco committed:
circle: Do not cache node_modules

* And this is what happens when someone makes a commit after 2am.

n-riesco committed:
ibmdb: Use electron-mocha

* Use electron-mocha so that we don't need to rebuild the native
  modules.

n-riesco committed:
ibmdb: Remove npm script `fix:ibm_db`

* Use `install:modules:electron` instead.

n-riesco committed:
circle: Rebuild native modules for electron

n-riesco committed:
test: Increase timeout of updateGrid unit test

n-riesco committed:
circle: Add extension RANDR to Xvfb

n-riesco committed:
Revert "circle: Add extension RANDR to Xvfb"

This reverts commit 2be5f851ee2dc46aed571f4dece778b1806af4d0.

n-riesco committed:
Revert "test: Increase timeout of updateGrid unit test"

This reverts commit 6bb710e8783ac14e16df8f411225124556ad3be9.

n-riesco committed:
test: Add debug messages

n-riesco committed:
test: Add log messages

n-riesco committed:
test: Increase timeout for PlotlyAPI

n-riesco committed:
Revert "test: Add log messages"

This reverts commit f7f2a3b884cd888b55efcebfa5ee61c2c31916af.

n-riesco committed:
Revert "test: Add debug messages"

This reverts commit a365b4ac1a24c6c9956a7ed832a76e64c88ebbdc.

n-riesco committed:
Revert "test: Increase timeout for PlotlyAPI"

This reverts commit 1cf840e6b0733f3299a9c347994f6f443cb63ac2.

n-riesco committed:
circle: Do not build at the test stage

* Removed `npm run build` from `npm run test`.

* Moved `npm run build` to the compile stage at CircleCI

n-riesco committed:
test: Skip updateGrid test

* Skipped updateGrid test for the time being, because it times out in
  CircleCI.

n-riesco commented:
Merged upstream https://github.com/plotly/plotly-database-connector/pull/163
